src/c/gr_context.cpp

```diff
@@ -74,110 +85,184 @@ void gr_context_reset_context(gr_context_t* context, uint32_t state) {
-    return SK_ONLY_GPU_RETURN(ToGrGLInterface(GrGLMakeAssembledGLESInterface(ctx, get).release()), nullptr);
+    return SK_ONLY_GPU(ToGrGLInterface(GrGLMakeAssembledGLESInterface(ctx, get).release()), nullptr);
 }
 
 void gr_glinterface_unref(const gr_glinterface_t* glInterface) {
     SK_ONLY_GPU(SkSafeUnref(AsGrGLInterface(glInterface)));
 }
 
 bool gr_glinterface_validate(const gr_glinterface_t* glInterface) {
-    return SK_ONLY_GPU_RETURN(AsGrGLInterface(glInterface)->validate(), false);
+    return SK_ONLY_GPU(AsGrGLInterface(glInterface)->validate(), false);
 }
 
 bool gr_glinterface_has_extension(const gr_glinterface_t* glInterface, const char* extension) {
-    return SK_ONLY_GPU_RETURN(AsGrGLInterface(glInterface)->hasExtension(extension), false);
+    return SK_ONLY_GPU(AsGrGLInterface(glInterface)->hasExtension(extension), false);
 }
 
 
+// GrVkInstance
+
+gr_vkinterface_t* gr_vkinterface_make(vk_getinstanceprocaddr_t* vkGetInstanceProcAddr,
+                                      vk_getdeviceprocaddr_t* vkGetDeviceProcAddr,
+                                      vk_instance_t* vkInstance,
+                                      vk_device_t* vkDevice,
+                                      uint32_t extensionFlags) {
+    SK_ONLY_VULKAN(
+        sk_sp<GrVkInterface> grVkInterface(new GrVkInterface(
+            GrVkInterface::GetInstanceProc(reinterpret_cast<PFN_vkVoidFunction(*)(VkInstance, const char*)>(vkGetInstanceProcAddr)),
+            GrVkInterface::GetDeviceProc(reinterpret_cast<PFN_vkVoidFunction(*)(VkDevice, const char*)>(vkGetDeviceProcAddr)),
+            reinterpret_cast<VkInstance>(vkInstance),
+            reinterpret_cast<VkDevice>(vkDevice),
+            extensionFlags
+        ));
+    )
+
+    return SK_ONLY_VULKAN(ToGrVkInterface(grVkInterface.release()), nullptr);
+}
+
+void gr_vkinterface_unref(gr_vkinterface_t* grVkInterface) {
+    SK_ONLY_VULKAN(SkSafeUnref(AsGrVkInterface(grVkInterface)));
+}
+
+bool gr_vkinterface_validate(const gr_vkinterface_t* grVkInterface, uint32_t extensionsFlags)
+{
+    return SK_ONLY_VULKAN(AsGrVkInterface(grVkInterface)->validate(extensionsFlags), false);
+}
+
+// GrVkBackendContext
+
+gr_vkbackendcontext_t* gr_vkbackendcontext_assemble(vk_instance_t* vkInstance,
+                                                    vk_physical_device_t* vkPhysicalDevice,
+                                                    vk_device_t* vkDevice,
+                                                    vk_queue_t* vkQueue,
+                                                    uint32_t graphicsQueueIndex,
+                                                    uint32_t minAPIVersion,
+                                                    uint32_t extensions,
+                                                    uint32_t features,
+                                                    gr_vkinterface_t* grVkInterface) {
+    SK_ONLY_VULKAN(
```

Comment:
Similar to https://github.com/mono/skia/pull/69/files#r346583773, I would rather this be simpler and the construction be done via properties. Also, use the delegate form to construct the actual instance - rather than setting each property here.

Reply:
Those are not really "properties", those parameters need to be coherent with the ones used to create the queue.